lib/data/external/GCP/GcpApis/copyObject.js

```diff
@@ -0,0 +1,122 @@
+        },
+        (token, next) => {
+            const uri = '/storage/v1' +
+                        `/b/${encodeURIComponent(sourceBucket)}` +
+                        `/o/${encodeURIComponent(sourceObject)}` +
+                        '/copyTo' +
+                        `/b/${encodeURIComponent(params.Bucket)}` +
+                        `/o/${encodeURIComponent(params.Key)}`;
+            request({
+                method: 'POST',
+                baseUrl: this.config.jsonEndpoint,
+                uri,
+                auth: { bearer: token } },
+            (err, resp, body) => {
+                if (err) {
+                    logHelper(logger, 'error',
+                        'copyObject: err in json method',
+                        errors.InternalError.customizeDescription(
+                            'json method copyObject failed'));
+                    return next(errors.InternalError
+                        .customizeDescription('err in JSON Request'));
+                }
+                if (resp.statusCode >= 300) {
+                    return next(
+                        new JsonError(resp.statusMessage, resp.statusCode));
+                }
+                let res;
+                try {
+                    res = body && typeof body === 'string' ?
+                        JSON.parse(body) : body;
+                } catch (error) { res = undefined; }
+                if (res && res.error && res.error.code >= 300) {
+                    return next(
+                        new JsonError(res.error.message, res.error.code));
+                }
+                return next(null, res);
+            });
+        },
+        (result, next) => {
+            // if metadata directive is REPLACE then perform a metadata update
+            // otherwise default to COPY
+            if (params.MetadataDirective === 'REPLACE') {
+                const updateParams = {
+                    Bucket: params.Bucket,
+                    Key: params.Key,
+                    Metadata: params.Metadata || {},
+                    VersionId: result.generation,
+                };
+                return this.updateMetadata(updateParams, (err, res) => {
+                    if (err) {
+                        return next(err);
+                    }
+                    const resObj = {
+                        CopyObjectResult: {
+                            ETag: res.etag,
+                            LastModified: res.updated,
+                        },
+                        VersionId: res.generation,
+                    };
+                    return next(null, resObj);
+                });
+            }
+            const resObj = {
+                CopyObjectResult: {
+                    ETag: result.etag,
+                    LastModified: result.updated,
+                },
+                VersionId: result.generation,
+            };
+            return next('SkipMetadata', resObj);
```

Comment:
In this method, it might be clearer to just return `next(null, result)` here and `next(null, res)` above. Then in the end of the waterfall construct the object once. That way we don't have to construct the object twice or send 'SkipMetadata' as an error.